### ISSUE TITLE
Fail2ban inconsistent log processing between servers

### ISSUE THREAD

**Issue by vmswtje** [fail2ban/fail2ban#2510]
### Environment:

- Fail2Ban version (including any possible distribution suffixes): 0.9.3-1
- OS, including release name/version: Ubuntu 16.04 Xenial
- [x] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [x] Some customizations were done to the configuration (provide details below is so)

### The issue:

Sometimes input-log-lines are ignored (not detected?) while fail2ban is running. Only when I restart fail2ban, the ipblocks will recover themself because fail2ban checks all logs again. 

It seems like Fail2ban misses some new lines/updates on the log, even though newer lines get detected. Because of this problem, some lines are missed.

#### Steps to reproduce

External (irrelevant) server writes fail2ban.log over the network using rsyslog. I've got 3 servers that receive those loglines, all using the same software and configuration, as far I can find out.
All receiving servers write the lines in a regular file using the rsyslog-server: /srv/log/log/fail2ban.log. I can confirm that the logs are the same.

Using 3 custom filters & jails, I distinguish 3 kinds of "levels" (1 hour ban, 1 day ban, 1 week ban) and they all use the that same log.

#### Expected behavior

All 3 fail2ban-installations interpret the logs the same and apply the same bans.

#### Observed behavior

The bans differ between the servers.
One of the servers (hosting17) blocks around 50, hosting16 around 30 en hosting14 around 25. There are 50 log lines, so there should be 50 blocks.

#### Any additional information

- All /srv/log/log/fail2ban.log-files are the same (so the input is the same)
- There are no leftover (shorewall) ipblocks
- I tried it multiple times
- I tried the :memory: database (in order to have fresh start every fail2ban restart)
- Even when waiting for additional time, many loglines gets missed and will never get processed (while newer lines are followed upon)
- In most cases the same servers have less ip-blocks. Like there's something wrong that causes the server to skip lines.
- The servers aren't really busy (enough cpu/memory/storage)
- I tried using gamin and polling, results are the same

**- When restarting fail2ban, all ipblocks are the same again (for some time, but when newer lines gets added, problem returns)**

### Configuration, dump and another helpful excerpts
**fail2ban-client -vvv -x start on the server that misses the most of the lines (in my last test)**
```INFO   Loading configs for fail2ban under /etc/fail2ban
DEBUG  Reading configs for fail2ban under /etc/fail2ban
DEBUG  Reading config files: /etc/fail2ban/fail2ban.conf
INFO     Loading files: ['/etc/fail2ban/fail2ban.conf']
INFO     Loading files: ['/etc/fail2ban/fail2ban.conf']
INFO   Using socket file /var/run/fail2ban/fail2ban.sock
INFO   Loading configs for jail under /etc/fail2ban
DEBUG  Reading configs for jail under /etc/fail2ban
DEBUG  Reading config files: /etc/fail2ban/jail.conf, /etc/fail2ban/jail.d/tf-fail2ban-dag.conf, /etc/fail2ban/jail.d/tf-fail2ban-uur.conf, /etc/fail2ban/jail.d/tf-fail2ban-week.conf
INFO     Loading files: ['/etc/fail2ban/jail.conf']
INFO     Loading files: ['/etc/fail2ban/paths-debian.conf']
INFO     Loading files: ['/etc/fail2ban/paths-common.conf']
INFO     Loading files: ['/etc/fail2ban/paths-overrides.local']
INFO     Loading files: ['/etc/fail2ban/jail.d/tf-fail2ban-dag.conf']
INFO     Loading files: ['/etc/fail2ban/jail.d/tf-fail2ban-uur.conf']
INFO     Loading files: ['/etc/fail2ban/jail.d/tf-fail2ban-week.conf']
INFO     Loading files: ['/etc/fail2ban/paths-common.conf', '/etc/fail2ban/paths-debian.conf', '/etc/fail2ban/jail.conf', '/etc/fail2ban/jail.d/tf-fail2ban-dag.conf', '/etc/fail2ban/jail.d/tf-fail2ban-uur.conf', '/etc/fail2ban/jail.d/tf-fail2ban-week.conf']
INFO   Loading configs for filter.d/tf-fail2ban-dag under /etc/fail2ban
DEBUG  Reading configs for filter.d/tf-fail2ban-dag under /etc/fail2ban
DEBUG  Reading config files: /etc/fail2ban/filter.d/tf-fail2ban-dag.conf
INFO     Loading files: ['/etc/fail2ban/filter.d/tf-fail2ban-dag.conf']
INFO     Loading files: ['/etc/fail2ban/filter.d/common.conf']
INFO     Loading files: ['/etc/fail2ban/filter.d/common.local']
INFO     Loading files: ['/etc/fail2ban/filter.d/common.conf', '/etc/fail2ban/filter.d/tf-fail2ban-dag.conf']
INFO   Loading configs for action.d/shorewall under /etc/fail2ban
DEBUG  Reading configs for action.d/shorewall under /etc/fail2ban
DEBUG  Reading config files: /etc/fail2ban/action.d/shorewall.conf
INFO     Loading files: ['/etc/fail2ban/action.d/shorewall.conf']
INFO     Loading files: ['/etc/fail2ban/action.d/shorewall.conf']
INFO   Loading configs for filter.d/tf-fail2ban-uur under /etc/fail2ban
DEBUG  Reading configs for filter.d/tf-fail2ban-uur under /etc/fail2ban
DEBUG  Reading config files: /etc/fail2ban/filter.d/tf-fail2ban-uur.conf
INFO     Loading files: ['/etc/fail2ban/filter.d/tf-fail2ban-uur.conf']
INFO     Loading files: ['/etc/fail2ban/filter.d/common.conf', '/etc/fail2ban/filter.d/tf-fail2ban-uur.conf']
INFO   Loading configs for filter.d/tf-fail2ban-week under /etc/fail2ban
DEBUG  Reading configs for filter.d/tf-fail2ban-week under /etc/fail2ban
DEBUG  Reading config files: /etc/fail2ban/filter.d/tf-fail2ban-week.conf
INFO     Loading files: ['/etc/fail2ban/filter.d/tf-fail2ban-week.conf']
INFO     Loading files: ['/etc/fail2ban/filter.d/common.conf', '/etc/fail2ban/filter.d/tf-fail2ban-week.conf']
INFO   [#         ] Waiting on the server...DEBUG  Starting '/usr/bin/fail2ban-server' with args ['fail2ban-server', '-s', '/var/run/fail2ban/fail2ban.sock', '-p', '/var/run/fail2ban/fail2ban.pid', '-x', '-b']
2019-08-22 16:47:08,942 fail2ban.server         [5403]: INFO    Starting Fail2ban v0.9.3
2019-08-22 16:47:08,942 fail2ban.server         [5403]: INFO    Starting in daemon mode
DEBUG  OK : 'pong'

DEBUG  OK : 'auto'
DEBUG  OK : 'INFO'
DEBUG  OK : '/var/log/fail2ban.log'
DEBUG  OK : ':memory:'
DEBUG  OK : 86400
DEBUG  OK : 'tf-fail2ban-dag'
DEBUG  OK : ['/srv/log/log/fail2ban.log']
DEBUG  OK : 86400
DEBUG  OK : 1
DEBUG  OK : 86400
DEBUG  OK : ''
DEBUG  OK : ['127.0.0.1/8']
DEBUG  OK : 'ISO-8859-1'
DEBUG  OK : 'warn'
DEBUG  OK : ['^(\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*| .*(?:\\[\\d+\\])?:\\s+)NOTICE\\s+\\[tfpostfix.*\\]\\s+Ban\\s+(?:::f{4,6}:)?(?P<host>[\\w\\-.^_]*\\w)\\s*$']
DEBUG  OK : 'shorewall'
DEBUG  OK : ''
DEBUG  OK : 'shorewall <blocktype> <ip>'
DEBUG  OK : ''
DEBUG  OK : 'shorewall allow <ip>'
DEBUG  OK : ''
DEBUG  OK : 'reject'
DEBUG  OK : 'tf-fail2ban-uur'
DEBUG  OK : ['/srv/log/log/fail2ban.log']
DEBUG  OK : 3600
DEBUG  OK : 1
DEBUG  OK : 3600
DEBUG  OK : ''
DEBUG  OK : ['127.0.0.1/8']
DEBUG  OK : 'ISO-8859-1'
DEBUG  OK : 'warn'
DEBUG  OK : ['^(\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*| .*(?:\\[\\d+\\])?:\\s+)NOTICE\\s+\\[shorewall\\]\\s+Ban\\s+(?:::f{4,6}:)?(?P<host>[\\w\\-.^_]*\\w)\\s*$']
DEBUG  OK : 'shorewall'
DEBUG  OK : ''
DEBUG  OK : 'shorewall <blocktype> <ip>'
DEBUG  OK : ''
DEBUG  OK : 'shorewall allow <ip>'
DEBUG  OK : ''
DEBUG  OK : 'reject'
DEBUG  OK : 'tf-fail2ban-week'
DEBUG  OK : ['/srv/log/log/fail2ban.log']
DEBUG  OK : 604800
DEBUG  OK : 1
DEBUG  OK : 604800
DEBUG  OK : ''
DEBUG  OK : ['127.0.0.1/8']
DEBUG  OK : 'ISO-8859-1'
DEBUG  OK : 'warn'
DEBUG  OK : ['^(\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*| .*(?:\\[\\d+\\])?:\\s+)NOTICE\\s+\\[recidiveweek\\]\\s+Ban\\s+(?:::f{4,6}:)?(?P<host>[\\w\\-.^_]*\\w)\\s*$']
DEBUG  OK : 'shorewall'
DEBUG  OK : ''
DEBUG  OK : 'shorewall <blocktype> <ip>'
DEBUG  OK : ''
DEBUG  OK : 'shorewall allow <ip>'
DEBUG  OK : ''
DEBUG  OK : 'reject'
DEBUG  OK : None
DEBUG  OK : None
DEBUG  OK : None
```

#### Any customizations done to /etc/fail2ban/ configuration
All have the same configuration, so I'll drop one of them here
```
root@hosting14:/etc/fail2ban/filter.d# cat tf-fail2ban-dag.conf
# Fail2ban custom filter
# Managed by Puppet

[Init]

[INCLUDES]
before = common.conf

[Definition]


failregex = ^(%(__prefix_line)s| .*%(__pid_re)s?:\s+)NOTICE\s+\[tfpostfix.*\]\s+Ban\s+<HOST>\s*$

ignoreregex =
root@hosting14:/etc/fail2ban/filter.d# cat tf-fail2ban-uur.conf
# Fail2ban custom filter
# Managed by Puppet

[Init]

[INCLUDES]
before = common.conf

[Definition]


failregex = ^(%(__prefix_line)s| .*%(__pid_re)s?:\s+)NOTICE\s+\[shorewall\]\s+Ban\s+<HOST>\s*$

ignoreregex =
root@hosting14:/etc/fail2ban/filter.d# cat tf-fail2ban-week.conf
# Fail2ban custom filter
# Managed by Puppet

[Init]

[INCLUDES]
before = common.conf

[Definition]


failregex = ^(%(__prefix_line)s| .*%(__pid_re)s?:\s+)NOTICE\s+\[recidiveweek\]\s+Ban\s+<HOST>\s*$

ignoreregex =

#
# Custom jail tf-fail2ban-dag
# Managed by Puppet
#

[tf-fail2ban-dag]
enabled  = true
action   = shorewall
filter   = tf-fail2ban-dag
logpath  = /srv/log/log/fail2ban.log
maxretry = 1
findtime     = 86400
bantime  = 86400
root@hosting14:/etc/fail2ban/jail.d# cat tf-fail2ban-uur.conf
#
# Custom jail tf-fail2ban-uur
# Managed by Puppet
#

[tf-fail2ban-uur]
enabled  = true
action   = shorewall
filter   = tf-fail2ban-uur
logpath  = /srv/log/log/fail2ban.log
maxretry = 1
findtime     = 3600
bantime  = 3600
root@hosting14:/etc/fail2ban/jail.d# cat tf-fail2ban-week.conf
#
# Custom jail tf-fail2ban-week
# Managed by Puppet
#

[tf-fail2ban-week]
enabled  = true
action   = shorewall
filter   = tf-fail2ban-week
logpath  = /srv/log/log/fail2ban.log
maxretry = 1
findtime     = 604800
bantime  = 604800
```

#### Relevant parts of /var/log/fail2ban.log file:
_preferably obtained while running fail2ban with `loglevel = 4`_

In my test, only 2 of the servers did this block, but one of them didn't see it
```
2019-08-22 16:53:53,159 fail2ban.filter         [5405]: INFO    [tf-fail2ban-dag] Found 104.148.70.53
2019-08-22 16:53:53,788 fail2ban.actions        [5405]: NOTICE  [tf-fail2ban-dag] Ban 104.148.70.53
```

#### Relevant lines from monitored log files in question:

(Caution: you are right, this is also a fail2ban-log, but this one is from another external server)
```
Aug 22 16:53:53 log01 fail2ban.actions        [24453]: NOTICE  [tfpostfixdirect] Ban 104.148.70.53
```

**Comment by sebres** [comment]
> In my test, only 2 of the servers did this block, but one of them didn't see it

Ok, you don't see `Ban ...`, but do you see `Found ...`?

1. In negative case, the filter is affected (or jail is a bit different? ignoreip? etc).
To be sure make dump (`fail2ban-client -d | grep tf-fail2ban-dag`) on 2 servers and compare the results.
Or check it from the server (which not find all entries) using `fail2ban-regex` how many entries it would match. 
Try `fail2ban-regex /srv/log/log/fail2ban.log /etc/fail2ban/filter.d/tf-fail2ban-dag.conf`
If it is 50, then for some reason fail2ban-server works different as fail2ban-regex.
If you see also fewer as 50 - something should be different in jail/filter configs.

2. If found (just no ban) - then the filter looks good, but there are also several reasons why it can be - see [FAQ in wiki, first answer](https://github.com/fail2ban/fail2ban/wiki/How-fail2ban-works).

3. Or simply the time-zones of both servers (or even its local time) are different and fail2ban assumes the messages are obsolete - too old regarding `findtime` (`found-time < now - findtime`), or even end-of ban is exceeded (`now > found-time + bantime`), so it ignores that.

**Comment by sebres** [comment]
> Only when I restart fail2ban, the ipblocks will recover themself because fail2ban checks all logs again.

Hmm... How it can be? If you don't have specified `head` option after logpath. fail2ban will scan only from last  position in the same log.

BTW. 0.9.3 is too old (I have no idea whether some of fixed errors could cause this situation, there are just too many).

**Comment by vmswtje** [comment]
`Ok, you don't see Ban ..., but do you see Found ...?`
No, nothing at all

`To be sure make dump (fail2ban-client -d | grep tf-fail2ban-dag) on 2 servers and compare the results.`
Same data, only difference is the ordering. Output:
```
['add', 'tf-fail2ban-dag', 'auto']
['set', 'tf-fail2ban-dag', 'ignorecommand', '']
['set', 'tf-fail2ban-dag', 'bantime', 86400]
['set', 'tf-fail2ban-dag', 'usedns', 'warn']
['set', 'tf-fail2ban-dag', 'addlogpath', '/srv/log/log/fail2ban.log', 'head']
['set', 'tf-fail2ban-dag', 'logencoding', 'auto']
['set', 'tf-fail2ban-dag', 'findtime', 86400]
['set', 'tf-fail2ban-dag', 'addignoreip', '127.0.0.1/8']
['set', 'tf-fail2ban-dag', 'maxretry', 1]
['set', 'tf-fail2ban-dag', 'addfailregex', '^(\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?\\S*(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*| .*(?:\\[\\d+\\])?:\\s+)NOTICE\\s+\\[tfpostfix.*\\]\\s+Ban\\s+<HOST>\\s*$']
['set', 'tf-fail2ban-dag', 'addaction', 'shorewall']
['set', 'tf-fail2ban-dag', 'action', 'shorewall', 'actioncheck', '']
['set', 'tf-fail2ban-dag', 'action', 'shorewall', 'actionstart', '']
['set', 'tf-fail2ban-dag', 'action', 'shorewall', 'actionunban', 'shorewall allow <ip>']
['set', 'tf-fail2ban-dag', 'action', 'shorewall', 'actionban', 'shorewall <blocktype> <ip>']
['set', 'tf-fail2ban-dag', 'action', 'shorewall', 'actionstop', '']
['set', 'tf-fail2ban-dag', 'action', 'shorewall', 'blocktype', 'reject']
['start', 'tf-fail2ban-dag']
```

`Try fail2ban-regex /srv/log/log/fail2ban.log /etc/fail2ban/filter.d/tf-fail2ban-dag.conf`

Exactly the same output on all 3 servers:

```Running tests
=============

Use   failregex filter file : tf-fail2ban-dag, basedir: /etc/fail2ban
Use         log file : /srv/log/log/fail2ban.log
Use         encoding : ISO-8859-1


Results
=======

Failregex: 1167 total
|-  #) [# of hits] regular expression
|   1) [1167] ^(\s*(<[^.]+\.[^.]+>)?\s*(?:\S+ )?(?:kernel: \[ *\d+\.\d+\] )?(?:@vserver_\S+ )?(?:(?:\[\d+\])?:\s+[\[\(]?\S*(?:\(\S+\))?[\]\)]?:?|[\[\(]?\S*(?:\(\S+\))?[\]\)]?:?(?:\[\d+\])?:?)?\s(?:\[ID \d+ \S+\])?\s*| .*(?:\[\d+\])?:\s+)NOTICE\s+\[tfpostfix.*\]\s+Ban\s+<HOST>\s*$
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [2809] (?:DAY )?MON Day 24hour:Minute:Second(?:\.Microseconds)?(?: Year)?
`-

Lines: 2809 lines, 0 ignored, 1167 matched, 1642 missed [processed in 0.30 sec]
Missed line(s): too many to print.  Use --print-all-missed to print all 1642 lines```

`Hmm... How it can be? If you don't have specified head option after logpath. fail2ban will scan only from last position in the same log.`

I just did `iptables -L -n | grep reject | wc -l` on all 3 servers. The one that's almost always right has 600 lines, the other two 167 resp. 133.
After restarting fail2ban and waiting for 10 minutes or so, the server of 167 lines did now ban 722 lines (I guess it will be more in about 20 minutes or so, because I see that only one jail has been processed). The server with 133 lines also banned 722 lines. So I guess they both missed many lines while monitoring, even though after restarting the processed it corrected the mistakes.

I think my config doesn't do tail but just processes the whole logs.

**Comment by vmswtje** [comment]
I circumvented the problem by using a new tactic. While debugging/trying to solve this problem, I removed fail2ban from the "nodes" (the three servers) and implemented ansible from the central fail2ban-server to send bans to shorewall on all nodes. The central server is an Ubuntu 18.04 (with fail2ban 0.10) so this also solves the "old package" problem.

My new tactic seems to work, so I'm happy with it. Thank you for your help!
